Add a config option to turn off comments/annotations site-wide

stevenday opened the pull request:
This adds a new config option `ENABLE_ANNOTATIONS` which defaults to `true`
but which allows you to turn off the commenting on InfoRequests if desired.

I've implemented this in two places - firstly in the "after actions" view, to
hide or show the comment link, and then on the `before_filter` for the actual
comment form action. Is that everywhere we offer the ability to comment?
(There are various open tickets about providing bulk annotations, or
redirecting straight to the annotation form after submission, but I couldn't
find anything like that actually in the code at the moment).

Closes #2955

garethrees commented:
Build failure looks like its because of https://github.com/mysociety/alaveteli/pull/2889#issuecomment-164815995 – I'll check it locally but should be safe to ignore.

garethrees commented:
Cheers! Looks good, but there are quite a few other places that annotations are mentioned (e.g. https://github.com/mysociety/alaveteli/blob/develop/app/controllers/request_controller.rb#L496) – probably need to cover these off.

stevenday commented:
For my future self: @garethrees was linking to https://github.com/mysociety/alaveteli/blob/208c1f766d623109c02c984d04e2be40b3999e3a/app/controllers/request_controller.rb#L496 as an example. It looks like (annoyingly) we'll have to create a whole bunch of new translation strings for when this is turned off :(

stevenday commented:
@garethrees - wow, that got complicated fast. 

I've been through and made every mention of annotations conditional on this new config, but it's a lot of changes. Before I start writing tests for all of these, I just want to check whether you and @crowbot will be happy to merge this?

stevenday commented:
Note: I haven't removed mention of annotations from the admin because I assumed that was unecessary.

Neither have I made any changes to the mailers that send emails about events (that might include annotations). I think that means there's is a possible "race condition" where you allow comments, then turn them off, and people get alerts about them inbetween, but I've assumed that that doesn't really matter.

I'm happy to revisit either of these assumptions if needs be.

garethrees commented on app/views/request/_after_actions.html.erb:
```diff
@@ -5,7 +5,7 @@
 
     <ul>
-      <% if @info_request.comments_allowed? %>
+      <% if AlaveteliConfiguration.enable_annotations && @info_request.comments_allowed? %>
```
You could bake `AlaveteliConfiguration.enable_annotations` in to `InfoRequest#comments_allowed?`.

stevenday replied:
I don't have a strong preference, but I would find that more confusing I think - it's a boolean field on the model, so the database and the ruby object would be telling different stories about what's happening. It also seems like giving the model too much extra knowledge about the application state?

garethrees replied:
Yeah, you're right there.

garethrees commented on app/views/request/_request_sent.html.erb:
```diff
@@ -40,14 +40,22 @@
-                'about your writing.',
-                :url => new_comment_url(:url_title => @info_request.url_title).html_safe) %>
-        </p>
+        <% AlaveteliConfiguration.enable_annotations %>
```
missing if

stevenday replied:
Thanks! I've fixed that up in 7d91c43

stevenday commented:
I've added specs (a83b588), fixed up https://github.com/mysociety/alaveteli/pull/2956#discussion_r57137166 in 7d91c43 and also fixed up a couple of other templates which needed some conditionals in 1e5f4cc

I note from this that alavetelitheme also contains copies of some of these templates which should (I guess?) have the conditionals too. Would you like me to make a matching PR there?

BenJam commented:
If there's any way this could be reviewed tomorrow it would be greatly appreciated. There's an inside chance that we will need to launch Infórmate Panamá April 1st.

garethrees commented on app/views/general/_advanced_search_tips.html.erb:
```diff
@@ -9,7 +9,9 @@
   <li><%= _('<strong><code>requested_from:home_office</code></strong> to search requests from the Home Office, typing the name as in the URL.')%></li>
   <li><%= _('<strong><code>requested_by:julian_todd</code></strong> to search requests made by Julian Todd, typing the name as in the URL.') %></li>
-  <li><%= _('<strong><code>commented_by:tony_bowden</code></strong> to search annotations made by Tony Bowden, typing the name as in the URL.')%></li>
+  <% AlaveteliConfiguration.enable_annotations %>
```
Missing `if`

stevenday replied:
Gah, I thought I'd caught that, fixed now.

garethrees commented on config/general.yml-example:
```diff
@@ -897,3 +897,14 @@ ENABLE_TWO_FACTOR_AUTH: false
+# If ENABLE_ANNOTATIONS is set to true, Alaveteli will include a link in the
+# "Things to do with this request" section to allow users to add extra
+# comments or information to it. This is a standard feature which is usually
+# turned on, but if you want to disable it, you can set this to false.
```
I think we should add a line like "We don't recommend disabling annotations after users have started making comments"

garethrees commented:
Its worth noting that existing comments are still shown, even if comments are disabled:

![screen shot 2016-03-31 at 09 10 58](https://cloud.githubusercontent.com/assets/282788/14170890/e6f3c5ae-f727-11e5-978e-371535680dc2.png)

I _think_ this is what we should do in that situation. If a site admin disables comments after the site has been live, it seems like it would make sense to preserve the history – they can always delete specific comments through the admin interface. 

If we did want to stop hiding them, we could apply the following diff:

```diff
diff --git a/app/views/request/_correspondence.html.erb b/app/views/request/_correspondence.html.erb
index a6263c8..8a80bdf 100644
--- a/app/views/request/_correspondence.html.erb
+++ b/app/views/request/_correspondence.html.erb
@@ -7,5 +7,7 @@
 <% when 'resent', 'followup_resent' %>
   <%= render :partial => 'request/resent_outgoing_correspondence', :locals => { :outgoing_message => info_request_event.outgoing_message, :info_request_event => info_request_event }%>
 <% when 'comment' %>
-  <%= render :partial => 'comment/single_comment', :locals => { :comment => info_request_event.comment } %>
+  <% if AlaveteliConfiguration.enable_annotations %>
+    <%= render :partial => 'comment/single_comment', :locals => { :comment => info_request_event.comment } %>
+  <% end %>
 <% end %>
```

garethrees commented:
@stevenday looks great. A couple of things left to do:

- [ ] [Small fixup](https://github.com/mysociety/alaveteli/pull/2956#discussion_r58020849) to the config documentation
- [ ] Add a `doc/CHANGES.md` entry
- [ ] Squash & Merge
- [ ] Make a PR to the [configuration documentation](http://alaveteli.org/docs/customising/config/)

Thanks!

stevenday commented:
@garethrees - thanks! I've done everything except squash and merge - happy for you to merge as much as you'd like. The documentation PR is here: https://github.com/mysociety/alaveteli/pull/3135,

I also made an alavetelitheme PR, which is here: https://github.com/mysociety/alavetelitheme/pull/76
